src/router_core/transfer.c

```diff
@@ -219,14 +217,32 @@ int qdr_link_process_deliveries(qdr_core_t *core, qdr_link_t *link, int credit)
                             dlv->where = QDR_DELIVERY_IN_UNSETTLED;
                             qd_log(core->log, QD_LOG_DEBUG, DLV_FMT"Delivery transfer:  qdr_link_process_deliveries: undelivered-list -> unsettled-list", DLV_ARGS(dlv));
                         }
+                    } else {
+                        //
+                        // This delivery is in the process of being transfered
+                        // to a different link.  Hack: the adaptor
+                        // deliver_handler has issued a new link-attached
+                        // action to the core, passing this dlv as the initial
```

Comment:
oh wow. I did not realize that in the TCP adaptor, the initial delivery could go to a new link on a totally DIFFERENT connection.

Reply:
Oh, I see - didn't realize that it could go to a separate connection.
Is the patch ok?  Should the comment be updated?

I was thinking perhaps we should try an alternative approach where we break this up into a two step process: first have the adaptor disassociate the dlv with its original link, then have the core handle adding it to the new link.

For example, we add a new API call for the adaptors to call during the "link deliver" callback that removes the delivery from the link (done while holding the work_lock of course).  The adaptor then owns the orphaned delivery until it can be passed to the core at a later time by calling the link_attach method, passing the initial delivery.

The link_process_deliveries loop would check for the MOVED_TO_NEW_LINK return code and simply update the original link's state and drop the delivery refcount.  It would be safe to assume the link->undelivered list has been updated at this point.

Thoughts?

Reply:
I don't think that's needed.  The MOVED_TO_NEW_LINK tells the core that the delivery has been referenced for use as the initial delivery on a different link.  The core (qdr_process_deliveries) should use this information to remove the delivery (complete or not) from the undelivered list because it is safely referenced elsewhere.  Just keep in mind that the delivery's new link may have already move it by the time MOVED_TO_NEW_LINK is seen.

There are two places where the removal from undelivered can happen:  in qdr_process_deliveries when MOVED_TO_NEW_LINK is returned, and in first_attach_CT when the delivery is supplied to a new link.  Exactly one of these must perform the switcheroo.

Reply:
To clarify... it already is a two step process.  Returning MOVED_TO... is the dissociation and first_attach(..., initial_dlv, ...) is the re-association.  The issue is that these steps can occur in either order.  In fact, they do:  HTTP2 does re-associate before dissociate and TCP does it the other way around.  In both cases, there are races between threads that can cause reversals.

Reply:
Do you think it would be a good idea for @mgoulish / @ChugR  to run the the tests for DISPATCH-1927 on this PR as well ?